File: cs/Refactor3/SRP_DI_Workshop/OrderProcessor.cs
using SRP_DI_Workshop.Domain.OrderOperations;
using SRP_DI_Workshop.ServiceInterface;

namespace SRP_DI_Workshop
{
    public class OrderProcessor
    {
        private readonly IResponseMessageFactory _responseMessageFactory;
        private readonly ILoggerService _loggerService;
        private readonly IDeduplicator _deduplicator;
        private readonly IOrderOperationFactory _orderOperationFactory;

        public OrderProcessor(
            IResponseMessageFactory responseMessageFactory,
            ILoggerService loggerService,
            IDeduplicator deduplicator,
            IOrderOperationFactory orderOperationFactory)
        {
            _responseMessageFactory = responseMessageFactory;
            _loggerService = loggerService;
            _deduplicator = deduplicator;
            _orderOperationFactory = orderOperationFactory;
        }

        public ResponseMessage ProcessOrder(RequestMessage reqMsg)
        {
            ResponseMessage responseToDuplicateMessage = _deduplicator.GetExistingResponseMessage(reqMsg.RequestId);

            if (responseToDuplicateMessage != null)
                return responseToDuplicateMessage;

            IOrderOperation orderOperation = _orderOperationFactory.GetOrderOperation(reqMsg.Operation);

            ResponseMessage response;

            if (orderOperation != null)
                response = orderOperation.ExecuteOperation(reqMsg);
            else
            {
                _loggerService.WriteLine("Received bad message: " + reqMsg.RequestId, "ProcessOrderError");
                response = _responseMessageFactory.CreateBadRequestResponseMessage(reqMsg);
            }

            _deduplicator.StoreResponseToMessage(reqMsg.RequestId, response);

            return response;
        }
    }
}

File: java/Refactor2/src/main/java/no/bouvet/solid/srpdip/OrderProcessor.java
package no.bouvet.solid.srpdip;

import no.bouvet.solid.srpdip.domain.orderoperations.OperationFactory;
import no.bouvet.solid.srpdip.domain.orderoperations.OrderOperation;
import no.bouvet.solid.srpdip.messageinterface.RequestMessage;
import no.bouvet.solid.srpdip.messageinterface.ResponseMessage;

public class OrderProcessor {

	private Deduplicator deduplicator = Deduplicator.factory.getInstance();
	private OperationFactory operationFactory = OperationFactory.factory.getInstance();

	public ResponseMessage processMessage(RequestMessage reqMsg) {
		ResponseMessage responseToDuplicateMessage = deduplicator.getExistingResponseMessage(reqMsg.getRequestId());

		if (responseToDuplicateMessage != null) {
			return responseToDuplicateMessage;
		}

		OrderOperation operation = operationFactory.getInstance(reqMsg.getOperation());
		ResponseMessage response = operation.execute(reqMsg);

		deduplicator.storeResponseToMessage(reqMsg.getRequestId(), response);

		return response;
	}
}
